File: 22.Automated-Testing-Unit-Integration-and-E2E/code/typescript/testing_with_a_real_database_6.ts
import { Client } from 'pg';
import { PostgreSqlContainer } from '@testcontainers/postgresql';

describe('UserRepo', () => {
    let container: any;
    let client: Client;

    beforeAll(async () => {
        container = await new PostgreSqlContainer().start();
        client = new Client({ connectionString: container.getConnectionUri() });
        await client.connect();
        await client.query('CREATE TABLE users (id TEXT, email TEXT)');
    });

    afterAll(async () => {
        await client.end();
        await container.stop();
    });

    it('saves and finds user', async () => {
        await client.query('INSERT INTO users (id, email) VALUES ($1, $2)', ['u1', 'a@x.com']);
        const res = await client.query('SELECT * FROM users WHERE id = $1', ['u1']);
        expect(res.rows[0].email).toBe('a@x.com');
    });
});

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/rust/testing_with_a_real_database_6.rs
// In rust, tests with database might use sqlx and testcontainers-rs.
// This is a pseudo-implementation to represent the concept.
/*
use sqlx::PgPool;

#[tokio::test]
async fn test_user_repo_save_and_find() {
    let pool = PgPool::connect("postgres://user:pass@localhost/db").await.unwrap();
    // Act
    sqlx::query!("INSERT INTO users (id, email) VALUES ($1, $2)", "u1", "a@x.com")
        .execute(&pool).await.unwrap();
    let row = sqlx::query!("SELECT email FROM users WHERE id = $1", "u1")
        .fetch_one(&pool).await.unwrap();
    // Assert
    assert_eq!(row.email, "a@x.com");
}
*/
